### PR TITLE
1st draft of crossword data post

### PR DESCRIPTION
**Pull request by railsagainstignorance**

### CODE REVIEW

File: _posts/2018-03-23-crossword-data-format.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+- (13,12) 22. ,23. Book a month to take over from glum donkey (4)
+- (15,12) 23. See 22 down (4)
+```
+The above example of the new crossword data format describes [Crossword No.15,800 Set by Julius on 2018/03/07](https://www.ft.com/__origami/service/image/v2/images/raw/ftcms:bdb7ea5a-1875-11e8-9e9c-25c814761640?source=crosswordsftcom),
+
+{% include figure-noimgservice.html src="https://www.ft.com/__origami/service/image/v2/images/raw/ftcms:bdb7ea5a-1875-11e8-9e9c-25c814761640?source=crosswordsftcom" alt="A example of the new crossword data format" width="100%" %}
+
+and drives [an interactive version](https://labs.ft.com/crosswords/2018/03/07/Puzzle-15800-dynamic.html).
+
+# The new format
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: Section titles should be `<h2>` not `<h1>`

File: _posts/2018-03-23-crossword-data-format.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+The crossword data format is a DSL, [a Domain Specific Language](https://en.wikipedia.org/wiki/Domain-specific_language), being an concise, clear, human-readable piece of text that specifies how to construct a simple crossword.
+
+It is written in [YAML](https://en.wikipedia.org/wiki/YAML). Or rather, nearly YAML. The line items for each clue need to be wrapped in quotes in order to accommodate colons and speech marks in the clue. Currently our implementation fails to take this into account, parsing the text directly and not as YAML.
+
+Blank lines and comments (starting with #) are ignored.
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: Feels like the characteristics should be a list

File: _posts/2018-03-23-crossword-data-format.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+There are several mandatory fields such as ‘name’ (which should be unique across all the crosswords), ‘author’, and ‘size’.
+
+The clues are grouped as lists named ‘across’ and ‘down’.
+
+Each clue has the form: - `(COORDINATES) IDS. Clue text (ANSWER)`
+
+COORDINATES are (across, down), with (1,1) being the top left square.
+
+The IDS identify each clue, and also tie together subsequent (comma-separated) clue ids for answers which straddle multiple positions on the grid. The subsequent IDS numbers might have an associated ‘down’ or ‘across’ if needed to disambiguate when a particular clue id has both aspects.
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: Variation between `IDS` & `ids` unclear; I'd probably settle for `IDs`

File: _posts/2018-03-23-crossword-data-format.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
+---
+layout: post
+title:  "Yet Another Crossword Data Format"
+date:   2018-03-23 10:30:30 +0100
+categories: Experiment
+teaser: Experiment
+excerpt: >
+  Defining a new data format for simple crosswords
+---
+When starting our interactive crossword project (link), we looked for an existing crossword data format, but not many were documented and widely used. Not wishing to create [Yet Another Standard](https://xkcd.com/927/), we went with [https://www.xwordinfo.com/JSON/](https://www.xwordinfo.com/JSON/) but it proved too error-prone, and forced a clumsy parse/build structure on the code. After several acrimonious intra-team discussions, we looked for ‘a better way’, and a new format was born.
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: Add `"_blank"` targets to all external links

File: _posts/2018-03-23-crossword-data-format.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+A little more effort (by unpacking the clues into a grid) is needed to ensure that
+
+* the answers fit within the stated size of the grid,
+* the clues are numbered appropriately,
+* and the answer letters agree at every intersection .
+
+We have several different incoming data formats for crosswords published by the FT, including some (undocumented) XML exported from a commercial crossword compiling tool, [CCW](https://www.crossword-compiler.com/), and a bespoke (but consistently-formatted) word doc. It has proved relatively simple to transform them into the new YAML format, and import that into our new crossword implementation.
+
+# Weaknesses
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: I would put this section as a subsection of `The new format`

File: _posts/2018-03-23-crossword-data-format.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+# Implementation / Infrastructure
+
+A couple of bugs/oversights have become apparent in our definition and implementation of this format, but let's pretend they are not a big issue for now.
+
+We have an in-browser [parser / validator / translator / editor / viewer](https://labs.ft.com/pages/crossword_dsl.html) which makes it easy to import and edit crosswords from a variety of sources. The code is ... horrible; a classic example of a hack which stayed around too long. The main fiddliness is in handling the multi-clue answers.
+
+The same parser is baked into an [Origami](https://origami.ft.com/) web [component](https://registry.origami.ft.com/components/o-crossword@1.8.2) which is used to power the current online test. The crossword data can be supplied as a url to a file, or directly as an attribute value.
+
+We use [Jekyll](https://jekyllrb.com/) to create the mini site for the crossword experiment, https://labs.ft.com/crosswords.
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: Missing crosswords link

File: _posts/2018-03-23-crossword-data-format.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+It is easy to parse and validate, at a basic level, to check that all the mandatory details and structure are present and correct.
+
+A little more effort (by unpacking the clues into a grid) is needed to ensure that
+
+* the answers fit within the stated size of the grid,
+* the clues are numbered appropriately,
+* and the answer letters agree at every intersection .
+
+We have several different incoming data formats for crosswords published by the FT, including some (undocumented) XML exported from a commercial crossword compiling tool, [CCW](https://www.crossword-compiler.com/), and a bespoke (but consistently-formatted) word doc. It has proved relatively simple to transform them into the new YAML format, and import that into our new crossword implementation.
```
Comment (Lily2point0) [inline _posts/2018-03-23-crossword-data-format.md]: "word document", or even "Word document"?

File: _posts/2018-03-23-crossword-data-format.md (hunk 2 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: possibly. makes it easier to mandate them if they are individual items

File: _posts/2018-03-23-crossword-data-format.md (hunk 4 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

File: _posts/2018-03-23-crossword-data-format.md (hunk 1 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

File: _posts/2018-03-23-crossword-data-format.md (hunk 3 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

File: _posts/2018-03-23-crossword-data-format.md (hunk 7 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

File: _posts/2018-03-23-crossword-data-format.md (hunk 5 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

File: _posts/2018-03-23-crossword-data-format.md (hunk 6 above)
Comment (railsagainstignorance) [reply _posts/2018-03-23-crossword-data-format.md]: done

**Pull request merged**